### ISSUE TITLE
URL::defaults() lost when controller uses route model binding

### ISSUE THREAD

**Issue by eballeste** [laravel/framework#34342]
- Laravel Version: 8.0.4
- PHP Version: 7.3.9
- Database Driver & Version: MySQL 5.7.26

### Description:
I set a default route parameter in a route middleware based on a Tenant model's slug. 

In a controller method that uses route model binding, a redirect to a named route throws: 
`Missing required parameters for [Route: dashboard] [URI: {tenant}]. UrlGenerationException`.

The same call in a method with no route model binding is able to redirect fine.

I tried to follow the stack trace but I lose it in a `1 unknown frame` right when the controller method is called.

### Steps To Reproduce:

`middleware tenant.domain`
```php
public function handle($request, Closure $next)
{
    URL::defaults(['tenant' => $request->tenant]);
    return $next($request);
}
```

`routes`
```php
Route::middleware('tenant.domain')->group(function () {
    Route::get('/{tenant}', [TenantController::class, 'index'])->name('dashboard');
    Route::get('/{tenant}/bound', [TenantController::class, 'bound']);
    Route::get('/{tenant}/unbound', [TenantController::class, 'unbound']);
});
```

`TenantController`
```php
public function index()
{
    return view('welcome');
}

public function bound(Tenant $tenant)
{
    return redirect(route('dashboard'));
}

public function unbound()
{
    return redirect(route('dashboard'));
}
```

`Tenant`
```php
public function getRouteKeyName()
{
    return 'slug';
}
```

`tenants migration`
```php
public function up()
{
    Schema::create('tenants', function (Blueprint $table) {
        $table->id();
        $table->string('name');
        $table->string('slug')->unique();
        $table->timestamps();
    });
}
```

`tenant seeder`
```php
public function run()
{
    Tenant::create([
        'name' => 'Test Tenant',
        'slug' => 'test',
    ]);
}
```
This is all a stripped down version, in real life the tenant parameter would be used as a subdomain.

I also set this up as a repo just in case.
https://github.com/eballeste/laravel-route-test

**Comment by driesvints** [comment]
So this happens because Laravel's `SubstituteBindings` middleware is run first, which replaces the substituted route binding with the model:

https://github.com/laravel/framework/blob/8.x/src/Illuminate/Routing/Middleware/SubstituteBindings.php#L39

To solve this problem of setting the defaults you need to prioritize your middleware before the `SubstituteBindings` one like this:

```php
<?php

namespace App\Http;

use Illuminate\Foundation\Http\Kernel as HttpKernel;

class Kernel extends HttpKernel
{
    ...

    /**
     * The priority-sorted list of middleware.
     *
     * This forces non-global middleware to always be in the given order.
     *
     * @var array
     */
    protected $middlewarePriority = [
        ...
        \App\Http\MiddlewareSetDefaultLocaleForUrls::class,
        \Illuminate\Routing\Middleware\SubstituteBindings::class,
        ...
    ];
}
```

This all is indeed not very obvious so I've made a PR to the docs to hopefully clarify this: https://github.com/laravel/docs/pull/6375

**Comment by eballeste** [comment]
Thanks for taking a look at my issue @driesvints!

The only thing about your solution is that it seems to force me to have to move my middleware from a route middleware to a global web middleware. I was hoping to contain this middleware to a small group of routes. 

After a lot of crawling from one method to another I figured out why the exception was being thrown:

`Illuminate\Routing\RouteUrlGenerator.php` has a `to` method that looks for missing required parameters and sets the uri with brackets in them for all of the missing parameters. When doing route model binding, the uri includes the model wrapped in brackets, instead of the default array `[{wildcard}, value]` array as part of the path which triggers the exception.

```php
if (preg_match('/\{.*?\}/', $uri)) {
    throw UrlGenerationException::forMissingParameters($route);
}
```

Knowing this, I updated my middleware so that it checks to see if the wildcard parameter is a slug(string) or model and if it's the model then I set the slug as the default value.

```php
public function handle($request, Closure $next)
 {
     $tenant_slug = is_string($request->tenant)
        ? $request->tenant : $request->tenant->slug;

     URL::defaults(['tenant' => $tenant_slug]);

     return $next($request);
 }
```
